File: ares-fabric-1.16/src/main/java/dev/tigr/ares/fabric/impl/modules/render/ESP.java
package dev.tigr.ares.fabric.impl.modules.render;

import dev.tigr.ares.core.feature.FriendManager;
import dev.tigr.ares.core.feature.module.Category;
import dev.tigr.ares.core.feature.module.Module;
import dev.tigr.ares.core.setting.Setting;
import dev.tigr.ares.core.setting.settings.BooleanSetting;
import dev.tigr.ares.core.setting.settings.EnumSetting;
import dev.tigr.ares.core.util.render.IRenderer;
import dev.tigr.ares.fabric.mixin.accessors.ShaderEffectAccessor;
import dev.tigr.ares.fabric.mixin.accessors.WorldRendererAccessor;
import dev.tigr.ares.fabric.utils.RenderUtils;
import dev.tigr.ares.fabric.utils.WorldUtils;
import net.minecraft.client.gl.GlUniform;
import net.minecraft.client.gl.PostProcessShader;
import net.minecraft.client.gl.ShaderEffect;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;

import java.util.List;

/**
 * @author Tigermouthbear
 */
@Module.Info(name = "ESP", description = "See outlines of players through walls", category = Category.RENDER)
public class ESP extends Module {
    private final Setting<Mode> mode = register(new EnumSetting<>("Mode", Mode.BOX));
    private final Setting<Color> color = register(new EnumSetting<>("Color", Color.RED)).setVisibility(() -> mode.getValue() == Mode.BOX);

    private final Setting<Boolean> players = register(new BooleanSetting("Players", true));
    private final Setting<Boolean> friends = register(new BooleanSetting("Friends", true)).setVisibility(players::getValue);
    private final Setting<Boolean> teammates = register(new BooleanSetting("Teammates", true)).setVisibility(players::getValue);
    private final Setting<Boolean> passive = register(new BooleanSetting("Passive", true));
    private final Setting<Boolean> hostile = register(new BooleanSetting("Hostile", true));
    private final Setting<Boolean> nametagged = register(new BooleanSetting("Nametagged", true));
    private final Setting<Boolean> bots = register(new BooleanSetting("Bots", false));

    @Override
    public void onRender3d() {
        if(MC.world.getPlayers().size() <= 0 || mode.getValue() != Mode.BOX) return;

        RenderUtils.prepare3d();

        for(Entity entity: WorldUtils.getTargets(players.getValue(), false, teammates.getValue(), passive.getValue(), hostile.getValue(), nametagged.getValue(), bots.getValue()))
            RenderUtils.renderEntityBoxNoPrepare(
                    entity,
                    new dev.tigr.ares.core.util.render.Color(0,0,0,0),
                    new dev.tigr.ares.core.util.render.Color(color.getValue().r, color.getValue().g, color.getValue().b, 1f),
                    2f,
                    0
            );

        if(friends.getValue()) renderFriends();

        RenderUtils.end3d();
    }

    private void renderFriends() {
        for(PlayerEntity player: MC.world.getPlayers()) {
            if(FriendManager.isFriend(player.getGameProfile().getName()) && !player.getUuid().equals(MC.player.getUuid()))
                RenderUtils.renderEntityBoxNoPrepare(player, new dev.tigr.ares.core.util.render.Color(0,0,0,0), IRenderer.rainbow());
        }
    }

    @Override
    public void onTick() {
        if(mode.getValue() != Mode.OUTLINE) return;

        ShaderEffect outlineShaderGroup = ((WorldRendererAccessor) MC.worldRenderer).getEntityOutlineShader();
        List<PostProcessShader> shaders = ((ShaderEffectAccessor) outlineShaderGroup).getPasses();

        shaders.forEach(shader -> {
            GlUniform outlineRadius = shader.getProgram().getUniformByName("Radius");
            if(outlineRadius != null) outlineRadius.set(4 / 5f);
        });

        for(Entity entity: MC.world.getEntities())
            entity.setGlowing(WorldUtils.isTarget(entity, players.getValue(), friends.getValue(), teammates.getValue(), passive.getValue(), hostile.getValue(), nametagged.getValue(), bots.getValue()));
    }

    @Override
    public void onDisable() {
        for(Entity entity: MC.world.getEntities()) entity.setGlowing(false);
    }

    @Override
    public String getInfo() {
        return mode.getValue().name();
    }

    enum Mode {BOX, OUTLINE}

    enum Color {
        RED(1, 0, 0),
        BLUE(0, 0, 1),
        BLACK(0, 0, 0),
        GREEN(0, 1, 0),
        WHITE(1, 1, 1);

        public float r;
        public float g;
        public float b;

        Color(float r, float g, float b) {
            this.r = r;
            this.g = g;
            this.b = b;
        }
    }
}

File: src/main/kotlin/org/aresclient/ares/impl/instrument/module/modules/render/ESP.kt
package org.aresclient.ares.impl.instrument.module.modules.render

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.VertexFormat
import net.minecraft.client.gl.SimpleFramebuffer
import net.minecraft.client.render.Frustum
import net.minecraft.client.render.OutlineVertexConsumerProvider
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityType
import net.minecraft.entity.LightningEntity
import net.minecraft.entity.boss.dragon.EnderDragonEntity
import net.minecraft.entity.decoration.DisplayEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.projectile.FishingBobberEntity
import net.minecraft.util.math.Box
import net.minecraft.util.math.MathHelper
import org.aresclient.ares.api.instruments.Module
import org.aresclient.ares.api.nrender.AresRenderPipelines
import org.aresclient.ares.api.nrender.world.WorldDrawer
import org.aresclient.ares.api.setting.settings.ColorSetting
import org.aresclient.ares.api.setting.settings.EnumSetting
import org.aresclient.ares.api.setting.settings.grouped.Group
import org.aresclient.ares.api.util.Color
import org.aresclient.ares.impl.instrument.module.components.render.esp.Chamlike
import org.aresclient.ares.impl.util.EntityUtil
import org.aresclient.ares.impl.util.EntityUtil.PlayerThreat
import org.aresclient.ares.impl.util.EntityUtil.playerThreat
import org.aresclient.ares.mixin.accessors.AccessWorldRenderer
import java.util.*
import kotlin.jvm.optionals.getOrNull

// TODO: FIX DEPTH ON OUTLINE ESP
object ESP: Module(Category.RENDER, "ESP", "See outlines of entities through walls") {
    enum class Mode { OUTLINE, BOX, CHAMLIKE }

    init {
        Chamlike
    }

    class EntityGroup(members: Set<Any> = emptySet()): Group<Any>(members) {
        companion object {
            fun create(title: String, members: Collection<Any>, mode: Mode, target: EntityUtil.Target, enabled: Boolean = false): EntityGroup {
                return EntityGroup(HashSet(members)).also {
                    it.title.value = title
                    it.mode.value = mode
                    it.lineColor.value = target.defaultColor
                    it.lineColor.isRainbow = target.defaultRainbow
                    it.fillColor.value = target.defaultColor.deriveAlpha(0.2f)
                    it.fillColor.isRainbow = target.defaultRainbow
                    it.enabled.value = enabled
                }
            }
        }

        val mode: EnumSetting<Mode> = addEnum("Mode", Mode.OUTLINE)
        val lineColor: ColorSetting = addColor("Line Color", Color.WHITE)
        val fillColor: ColorSetting = addColor("Fill Color", Color.WHITE.deriveAlpha(0.2f)).setVisibility { mode.value != Mode.OUTLINE } as ColorSetting
    }

    private val entities = settings.addGrouped("Entities", arrayListOf(
        EntityGroup.create("Friends", listOf(PlayerThreat.FRIEND), Mode.OUTLINE, PlayerThreat.FRIEND, enabled = true),
        EntityGroup.create("Players", listOf(PlayerThreat.HOSTILE), Mode.OUTLINE, PlayerThreat.HOSTILE, enabled = true),
        EntityGroup.create("Crystals", listOf(EntityType.END_CRYSTAL), Mode.CHAMLIKE, EntityUtil.TargetType.END_CRYSTAL, enabled = true),
        EntityGroup.create("Monsters", EntityUtil.EntityTypes.monster, Mode.OUTLINE, EntityUtil.TargetType.HOSTILE),
        EntityGroup.create("Animals", EntityUtil.EntityTypes.animal, Mode.OUTLINE, EntityUtil.TargetType.PASSIVE),
        EntityGroup.create("Items", listOf(EntityType.ITEM), Mode.OUTLINE, EntityUtil.TargetType.ITEM, enabled = true),
        EntityGroup.create("Misc", EntityUtil.EntityTypes.miscellaneous.filter { it != EntityType.END_CRYSTAL && it != EntityType.ITEM }, Mode.OUTLINE, EntityUtil.TargetType.OTHER)
    ), EntityUtil.EntityTypes.possibles, { EntityGroup() })

    private val entitiesCache = hashMapOf<Any, EntityGroup?>()
    private var shouldRenderOutlineCache: Boolean? = null

    override fun onTick() {
        entitiesCache.clear()
        shouldRenderOutlineCache = null
    }

    override fun onRenderWorld(drawer: WorldDrawer, delta: Float) {
        Chamlike.onRenderWorld(drawer, delta)
        if(entities.none { it.enabled.value && it.mode.value == Mode.BOX }) return

        val frustum = Frustum(MC.worldRenderer.capturedFrustum ?: (MC.worldRenderer as AccessWorldRenderer).frustum)
        WORLD.entities?.forEach { entity ->
            if(entity == SELF) return@forEach

            val group = getEntityGroup(entity) ?: return@forEach
            if(!group.enabled.value || group.mode.value != Mode.BOX || entity.shouldCull(frustum)) return@forEach

            val box = entity.getInterpolatedBoundingBox(delta)

            drawer.fillBox(box, group.fillColor.value)
            drawer.outlineBox(box, group.lineColor.value, 2f)
        }
    }

    fun getEntityColor(entity: Entity): Color = getEntityGroup(entity)?.lineColor?.value ?: Color.COLORLESS

    fun shouldRenderOutline(): Boolean {
        if(!isEnabled()) return false
        if(shouldRenderOutlineCache != null) return shouldRenderOutlineCache!!
        shouldRenderOutlineCache = WORLD.entities.any { entity -> getEntityGroup(entity)?.let { it.enabled.value && it.mode.value == Mode.OUTLINE } == true }
        return shouldRenderOutlineCache!!
    }

    fun shouldRenderOutline(entity: Entity) = shouldRenderOutline() && getEntityGroup(entity)?.let { it.enabled.value && it.mode.value == Mode.OUTLINE } == true

    fun getEntityGroup(entity: Entity): EntityGroup? {
        val type = if(entity is PlayerEntity) entity.playerThreat else entity.type as Any
        return entitiesCache.getOrPut(type) { entities.find(type).getOrNull() }
    }

    private fun <T: Entity> T.shouldCull(frustum: Frustum): Boolean {
        if(this is DisplayEntity && !shouldRender()) return true
        return when(this) {
            is EnderDragonEntity, is FishingBobberEntity, is LightningEntity -> false
            else -> !frustum.isVisible(boundingBox.expand(0.5))
        }
    }

    private fun Entity.getInterpolatedBoundingBox(delta: Float): Box {
        val camera = MC.gameRenderer.camera.pos
        val x = MathHelper.lerp(delta.toDouble(), lastRenderX, x) - x - camera.x
        val y = MathHelper.lerp(delta.toDouble(), lastRenderY, y) - y - camera.y
        val z = MathHelper.lerp(delta.toDouble(), lastRenderZ, z) - z - camera.z
        return Box(
            boundingBox.minX + x, boundingBox.minY + y, boundingBox.minZ + z,
            boundingBox.maxX + x, boundingBox.maxY + y, boundingBox.maxZ + z
        )
    }

    // see MixinWorldRenderer.java
    object Outliner {
        val framebuffer = SimpleFramebuffer("Ares Outline Framebuffer", MC.framebuffer.textureWidth, MC.framebuffer.textureHeight, true)
        val vertexConsumerProvider = OutlineVertexConsumerProvider(MC.bufferBuilders.entityVertexConsumers)

        fun setColor(color: Color) {
            vertexConsumerProvider.setColor((color.red * 255).toInt(), (color.green * 255).toInt(), (color.blue * 255).toInt(), 255) // TODO: ALPHA??
        }

        fun blit() {
            RenderSystem.assertOnRenderThread()
            val shapeIndexBuffer = RenderSystem.getSequentialBuffer(VertexFormat.DrawMode.QUADS)
            val gpuBuffer = shapeIndexBuffer.getIndexBuffer(6)
            val gpuBuffer2 = RenderSystem.getQuadVertexBuffer()
            RenderSystem.getDevice().createCommandEncoder().createRenderPass(MC.framebuffer.colorAttachment, OptionalInt.empty())
            .use { renderPass ->
                renderPass.setPipeline(AresRenderPipelines.OUTLINE)
                renderPass.setVertexBuffer(0, gpuBuffer2)
                renderPass.setIndexBuffer(gpuBuffer, shapeIndexBuffer.indexType)
                renderPass.bindSampler("theTexture", framebuffer.colorAttachment)
                renderPass.setUniform("viewportSize", MC.framebuffer.textureWidth.toFloat(), MC.framebuffer.textureHeight.toFloat())
                renderPass.setUniform("lineWeight", 1f)
                renderPass.drawIndexed(0, 6)
            }
            RenderSystem.getDevice().createCommandEncoder().clearColorAndDepthTextures(framebuffer.colorAttachment, 0, framebuffer.depthAttachment, 0.0)
        }
    }
}
